File: array_absurdity/array_absurdity.rb
# ARRAY ABSURDITY
# CHALLENGE DESCRIPTION:

# Imagine we have an immutable array of size N which we know to be filled with integers ranging from 0 to N-2, inclusive. Suppose we know that the array contains exactly one duplicated entry and that duplicate appears exactly twice. Find the duplicated entry. (For bonus points, ensure your solution has constant space and time proportional to N)

# INPUT SAMPLE:

# Your program should accept as its first argument a path to a filename. Each line in this file is one test case. Ignore all empty lines. Each line begins with a positive integer(N) i.e. the size of the array, then a semicolon followed by a comma separated list of positive numbers ranging from 0 to N-2, inclusive. i.e eg.

# 5;0,1,2,3,0
# 20;0,1,10,3,2,4,5,7,6,8,11,9,15,12,13,4,16,18,17,14
# OUTPUT SAMPLE:

# Print out the duplicated entry, each one on a new line eg

# 0
# 4

def solution line
  arr = line.split(';')[1].split ','
  arr.find { |i| arr.count(i) == 2 }
end

def init file
  IO.readlines(file).each { |line| puts solution line.chomp }
end

init ARGV[0]

File: array_absurdity/array_absurdity.js
function solution(line) {
    'use strict';
    var parts = line.split(';'),
        n = parseInt(parts[0], 10),
        arr = parts[1].split(',');
    for (var i = 0; i < n; i++) {
        var count = 0,
            current = arr[i];
        for (var j = 0; j < n; j++) {
            var current_comparison = arr[j];
            if (current === current_comparison) {
                count += 1;
            }
        }
        if (count > 1) {
            return current;
        }
    }
}

module.exports.solution = solution;

var fs = require('fs');
function init(file) {
    'use strict';
    fs.readFile(file, 'utf-8', function (err, data) {
        // prep the file data
        if (err) throw err;
        var lines = data.split('\n');
        lines.pop();

        // feed to solution line by line
        lines.forEach(function(line) {
        console.log(solution(line));
        });
    });
}

init(process.argv[2]);
